sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClient.cs

```diff
@@ -659,604 +600,482 @@ protected EventProcessorClient()
-                    }
-                    catch (EventHubsException ex)
-                    {
-                        var partitionId = Resources.OperationClaimOwnership.Equals(ex.GetFailureOperation()) ?
-                            ex.GetFailureData<string>() :
-                            null;
-                        var errorEventArgs = new ProcessErrorEventArgs(partitionId, ex.Message, ex.InnerException ?? ex, cancellationToken);
-                        _ = OnProcessErrorAsync(errorEventArgs);
+                        context ??= new ProcessorPartitionContext(partition.PartitionId, () => ReadLastEnqueuedEventProperties(partition.PartitionId));
+                        eventArgs = new ProcessEventArgs(context, eventData, updateToken => UpdateCheckpointAsync(eventData, context, updateToken), cancellationToken);
+
+                        await _processEventAsync(eventArgs).ConfigureAwait(false);
                     }
-                    catch (Exception ex)
+                    catch (Exception ex) when (!(ex is TaskCanceledException))
                     {
-                        Logger.LoadBalancingTaskError(Identifier, ex.Message);
-                        var errorEventArgs = new ProcessErrorEventArgs(null, string.Empty, ex, cancellationToken);
-                        _ = OnProcessErrorAsync(errorEventArgs);
-                    }
+                        // This exception is not surfaced to the error handler or bubbled, as the entire batch must be
+                        // processed or events will be lost.  Preserve the exceptions, should any occur.
 
-                    if (claimedOwnership != default)
-                    {
-                        await StartPartitionProcessingAsync(claimedOwnership.PartitionId, cancellationToken).ConfigureAwait(false);
+                        Logger.EventBatchProcessingError(partition.PartitionId, Identifier, EventHubName, ConsumerGroup, ex.Message);
+
+                        caughtExceptions ??= new List<Exception>();
+                        caughtExceptions.Add(ex);
                     }
                 }
 
-                // Some previously owned partitions might have had their ownership expired or might have been stolen, so we need to stop
-                // the processing tasks we don't need anymore.
-
-                await Task.WhenAll(ActivePartitionProcessors.Keys
-                    .Except(LoadBalancer.OwnedPartitionIds)
-                    .Select(partitionId => StopPartitionProcessingIfRunningAsync(partitionId, ProcessingStoppedReason.OwnershipLost, cancellationToken)))
-                    .ConfigureAwait(false);
+                // If the event batch was empty, then dispatch to the handler; the base class will ensure that empty batches
+                // are requested and will not invoke this method should empties not be sent to the handler.
```

Comment:
Do we actually have a knob for this behavior (where OnProcessingEventBatchAsync is not called with an empty enumerable in the case of a timeout) in the base class yet?

Reply:
Yup;  if the `MaximumWaitTime` is not set as part of the options (see [this doc comment](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/eventhub/Azure.Messaging.EventHubs.Processor/src/EventProcessorClientOptions.cs#L57), we do not invoke the handler with empties.   It gets filtered out at the base class.   Worth noting that the use of maximum wait time in this pattern is consistent across all clients and means of reading events.

Reply:
Gotcha - was thinking this implied there was going to be an additional knob that controlled if a empty batch was delivered in the timeout case when we had gotten zero events after the timeout, vs always delivering it as we've done today.